### DISCUSSION TITLE
Recommended solution for `T | None` type with `T = None` validation

### DISCUSSION THREAD

**Discussion by joaohaas** [pydantic/pydantic#6784] [Question]
Hi! I know there's been a lot (at least 5 issues/PRs) of discussion around this topic in the past, but that's exactly why I'm a bit confused as to which approach is preferred.

Is there a 'standard' way to get the behavior of this code:
```py
class A(BaseModel):
    b: str = None

A()
A(b='x')
try:
    A(b=None)
except ValidationError as e:
    print(e)
    """
    1 validation error for A
    b
      Input should be a valid string [type=string_type, input_value=None, input_type=NoneType]
    """
```
while using the correct type of `str | None`/`Optional[str]`?

My struct has at least 100 fields, so using field validators, although doable, would be annoying. I've seen mentions of `Unrequired[T]`, `T = Undefined`, `Field` with a custom attr, and `class Config` config, but I think none of these made it into pydantic. I guess I could also write the `Unrequired` wrapper myself, but I would prefer to use something more 'standard' if that's a thing.

**Comment by ImogenBits** [comment]
Do you mean a field that has a default value of None but you can't otherwise assign it None? If so you can do that like this:
```py
T = TypeVar("T")

def not_none(val: T) -> T:
    if val is None:
        raise ValueError("Value is None")
    return val

NotNone = Annotated[T, AfterValidator(not_none)]
OptionalButNotNone = NotNone[T | None]

class Example(BaseModel):
    attr: OptionalButNotNone[str] = None
```
If you don't care about IDE support you can even add the default to the annotation like this:
```py

NotNoneDefaultNone = Annotated[T, AfterValidator(not_none), Field(default=None)]
```
and then define them with just the type annotation
```py
class Example(BaseModel):
    attr: OptionalButNotNone[str]
```

**Comment by joaohaas** [comment]
Yeah, that's what I meant by possibly implementing the wrapper myself. Guess that's what I'll end up doing, thx.
